src/gwt/src/org/rstudio/studio/client/common/mathjax/MathJax.java

```diff
@@ -0,0 +1,445 @@
+         {
+         }
+      };
+      
+      ChunkOutputWidget outputWidget = new ChunkOutputWidget(
+            StringUtil.makeRandomId(8),
+            RmdChunkOptions.create(),
+            ChunkOutputWidget.EXPANDED,
+            host);
+      
+      outputWidget.setRootWidget(panel);
+
+      PinnedLineWidget plWidget = new PinnedLineWidget(
+            "mathjax",
+            docDisplay_,
+            outputWidget,
+            row,
+            null,
+            null);
+
+      cowToPlwMap_.put(outputWidget, plWidget);
+      return plWidget.getLineWidget();
+   }
+   
+   private void resetRenderState()
+   {
+      if (anchor_ != null)
+      {
+         anchor_.detach();
+         anchor_ = null;
+      }
+      
+      if (cursorChangedHandler_ != null)
+      {
+         cursorChangedHandler_.removeHandler();
+         cursorChangedHandler_ = null;
+      }
+      
+      lastRenderedText_ = "";
+   }
+   
+   private void renderPopup(final String text)
+   {
+      // no need to re-render if text hasn't changed or is empty
+      if (text.equals(lastRenderedText_))
+         return;
+      
+      // if empty, hide popup
+      if (text.isEmpty())
+      {
+         popup_.hide();
+         resetRenderState();
+         return;
+      }
+      
+      // no need to re-position popup if already showing;
+      // just typeset
+      if (popup_.isShowing())
+      {
+         mathjaxTypeset(popup_.getContentElement(), text);
+         return;
+      }
+      
+      // attach popup to DOM but render offscreen
+      popup_.setPopupPosition(100000, 100000);
+      popup_.show();
+      
+      // typeset and position after typesetting finished
+      mathjaxTypeset(popup_.getContentElement(), text, onPopupTypeset_);
+   }
+   
+   private void endRender()
+   {
+      resetRenderState();
+      popup_.hide();
+   }
+   
+   private void onMathJaxTypesetCompleted(String text,
+                                          boolean error,
+                                          Object commandObject)
+   {
+      if (commandObject != null && commandObject instanceof MathJaxTypesetCallback)
+      {
+         MathJaxTypesetCallback callback = (MathJaxTypesetCallback) commandObject;
+         callback.onMathJaxTypesetComplete(error);
+      }
+      
+      if (error) return;
+      lastRenderedText_ = text;
+   }
+   
+   private final void mathjaxTypeset(Element el, String currentText)
+   {
+      mathjaxTypeset(el, currentText, null);
+   }
+   
+   private final native void mathjaxTypeset(Element el,
+                                            String currentText,
+                                            Object command)
+   /*-{
+      var MathJax = $wnd.MathJax;
+      
+      // save last rendered text
+      var jax = MathJax.Hub.getAllJax(el)[0];
+      var lastRenderedText = jax && jax.originalText || "";
+      
+      // update text in element
+      el.innerText = currentText;
```

Comment:
Do we need to sanitize the input here when assigning to `innerText`? Or should we prefer a different method for setting text in this element? (since this is effectively user-supplied text within that chunk)